helix-core/src/main/java/org/apache/helix/task/TaskUtil.java

```diff
@@ -989,47 +988,65 @@ public static void purgeExpiredJobs(String workflow, WorkflowConfig workflowConf
     }
     long purgeInterval = workflowConfig.getJobPurgeInterval();
     long currentTime = System.currentTimeMillis();
-    final Set<String> expiredJobs = Sets.newHashSet();
+    Set<String> expiredJobs = Sets.newHashSet();
+    Set<String> jobsWithoutConfig = Sets.newHashSet();
     if (purgeInterval > 0 && workflowContext.getLastJobPurgeTime() + purgeInterval <= currentTime) {
-      expiredJobs.addAll(TaskUtil.getExpiredJobs(manager.getHelixDataAccessor(),
-          manager.getHelixPropertyStore(), workflowConfig, workflowContext));
-      if (expiredJobs.isEmpty()) {
+      TaskUtil.getExpiredJobs(manager.getHelixDataAccessor(),
+          manager.getHelixPropertyStore(), workflowConfig, workflowContext, expiredJobs, jobsWithoutConfig);
+
+      if (expiredJobs.isEmpty() && jobsWithoutConfig.isEmpty()) {
```

Comment:
I think instead of doing this, we should just deal with two things separately.

Reply:
Fixed.